### PR TITLE
[BACKPORT] Fix Spark CLI, tests, and build tools

### PR DESCRIPTION
**Pull request by samvantran**
Resolves: [DCOS-37965](https://jira.mesosphere.com/browse/DCOS-37965)

This PR backports Nick's PRs: #291 & #297 which includes **a lot** of tooling updates/CLI fixes. I did however revert quota related changes as they are outside the scope of a package release.

Most of the files changed are go vendor packages so feel free to skip those but worth looking at Makefile (in order to build distributions) and main/submit_builder.go/tests.

At the end of the day we finally get to: 
```
[23:04:59]	[Step 4/8] === RUN   TestCliTestSuite/TestScoptAppArgs
[23:04:59]	[Step 4/8] --- PASS: TestCliTestSuite (0.01s)
[23:04:59]	[Step 4/8]     --- PASS: TestCliTestSuite/TestCleanUpSubmitArgs (0.00s)
[23:04:59]	[Step 4/8]     --- PASS: TestCliTestSuite/TestPayloadCustomImageNoExecutor (0.00s)
[23:04:59]	[Step 4/8]     --- PASS: TestCliTestSuite/TestPayloadSimple (0.00s)
[23:04:59]	[Step 4/8]     --- PASS: TestCliTestSuite/TestPayloadWithSecret (0.00s)
[23:04:59]	[Step 4/8]     --- PASS: TestCliTestSuite/TestSaslSecret (0.00s)
[23:04:59]	[Step 4/8]     --- PASS: TestCliTestSuite/TestScoptAppArgs (0.00s)
[23:04:59]	[Step 4/8] PASS
[23:04:59]	[Step 4/8] ok  	github.com/mesosphere/build/cli/dcos-spark	0.019s
```
and 
```
[00:09:00]	[Step 7/8] ================== 18 passed, 5 deselected in 2402.26 seconds ==================
```

### CODE REVIEW

File: cli/dcos-spark/main.go [hunk 1]
```diff
@@ -186,9 +189,9 @@ func handleCommands(app *kingpin.Application) {
 
 	log := app.Command("log", "Retrieves a log file from a submitted Spark job").Action(cmd.runLog)
 	log.Flag("follow", "Dynamically update the log").BoolVar(&cmd.logFollow)
-	log.Flag("lines_count", "Print the last N lines.").
+	log.Flag("lines_count", "Print the last N lines").
 		Default("10").UintVar(&cmd.logLines) //TODO "lines"?
-	log.Flag("file", "Specify the sandbox file to print.").
+	log.Flag("file", "Specify the sandbox file to print").
 		Default("stdout").StringVar(&cmd.logFile)
 	log.Arg("submission-id", "The ID of the Spark job").Required().StringVar(&cmd.submissionId)
```
[the remark this answers is not in the document]
Comment (elezar) [reply cli/dcos-spark/main.go]: Good point. It's an `Arg` not a `Flag`. Then please keep the change to better align with `master`.

Looking at the help text:
```
secret <secret_path>
    Make a shared secret, used for RPC authentication
```

File: spark-testing/spark_utils.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        sdk_security.delete_service_account(service_account, secret)
+        LOGGER.info('Finished cleaning up strict-mode security for Spark')
+
+    try:
+        if sdk_utils.is_strict_mode():
+            setup_security()
+        yield
+    finally:
+        if sdk_utils.is_strict_mode():
+            cleanup_security()
```
Comment (elezar) [inline spark-testing/spark_utils.py]: May not be your fault, but we're missing a trailing newline here.

**Comment by samvantran** [comment]
@elezar yup, looks like adding #350 fixed up the strict builds (but a TC template change was also required)

File: cli/dcos-spark/main.go [hunk 3]
```diff
@@ -23,22 +23,25 @@ type SparkCommand struct {
 	submitArgs        string
 	submitDockerImage string
 	submitEnv         map[string]string
-	secretPath		  string
-
+	secretPath 		  string
```
Comment (elezar) [inline cli/dcos-spark/main.go]: I just ran `gofmt -w` on this file locally, and this fixes the formatting issues in this block and doesn't change anything else. I would recommend doing that.

File: cli/dcos-spark/main.go [hunk 4]
```diff
@@ -197,7 +200,7 @@ func handleCommands(app *kingpin.Application) {
 
 	secret := app.Command("secret", "Make a shared secret, used for RPC authentication").
 		Action(cmd.runGenerateSecret)
-	secret.Arg("secret_path", "path and name for the secret").Required().StringVar(&cmd.secretPath)
+	secret.Arg("secret-path", "path and name for the secret").Required().StringVar(&cmd.secretPath)
```
Comment (elezar) [inline cli/dcos-spark/main.go]: Thanks!

File: cli/dcos-spark/submit_builder.go [hunk 5]
```diff
@@ -248,7 +248,7 @@ func parseApplicationFile(args *sparkArgs) error {
 		if args.mainClass != "" {
 			return errors.New("Can only specify main class when using a Scala or Java Spark application")
 		}
-		log.Printf("Parsing application as R job")
+		client.PrintMessage("Parsing application as R job")
```
Comment (elezar) [inline cli/dcos-spark/submit_builder.go]: Running `gofmt -w submit_builder.go` changes a single line. Let's run that too.

File: cli/dcos-spark/submit_builder_test.go [hunk 6]
```diff
@@ -0,0 +1,275 @@
+package main
```
Comment (elezar) [inline cli/dcos-spark/submit_builder_test.go]: Since this file is moving anyway, let's run `gofmt -w` on it too.

File: conf/spark-env.sh [hunk 7]
```diff
@@ -69,7 +70,22 @@ if [[ -n "${KRB5CONF}" ]]; then
     echo "spark-env: Copying krb config from $KRB5CONF to /etc/" >&2
     echo "${KRB5CONF}" | ${BASE64_D} > /etc/krb5.conf
 else
-    echo "spark-env: No SPARK_MESOS_KRB5_CONF_BASE64 decoded" >&2
+    echo "spark-env: No SPARK_MESOS_KRB5_CONF_BASE64 to decode" >&2
+fi
+
+if [ -n "${STATSD_UDP_HOST}" ] && [ -n "${STATSD_UDP_PORT}" ]; then
```
Comment (elezar) [inline conf/spark-env.sh]: This change here adds metrics too. Is that ready for this release?

File: universe/config.json [hunk 8]
```diff
@@ -7,7 +7,7 @@
             "properties": {
                 "name": {
                     "default": "spark",
-                    "description": "The Spark Dispatcher will register with Mesos with this as a framework name.  This service will be available at http://<dcos_url>/service/<name>/",
+                    "description": "The Spark Dispatcher will register with Mesos with this as a framework name.  This service will be available at {{https-protocol}}<dcos-url>/service/<name>/",
```
Comment (elezar) [inline universe/config.json]: Nit: I know this is probably taken directly from the other change, but I don't see what value this templated value adds here.

File: conf/spark-env.sh (hunk 7 above)
Comment (samvantran) [reply conf/spark-env.sh]: Yes, product signed off and Nick created an upstream PR to update metrics too: https://github.com/apache/spark/pull/21516

File: universe/config.json (hunk 8 above)
Comment (samvantran) [reply universe/config.json]: /shrugs... it's in master though, we should change it there rather than here?

File: cli/dcos-spark/main.go (hunk 3 above)
Comment (samvantran) [reply cli/dcos-spark/main.go]: Ah yes! `gofmt` ... I'm still a golang newb so I forget about this but I do love it. I used to get into arguments with other devs over formatting whitespace in Scala/Java and here go embraces it. A++++

File: universe/config.json (hunk 8 above)
Comment (nickbp) [reply universe/config.json]: This was done to avoid errors with the package linter. It detects misplaced http:// URLs in packaging

**Comment by nickbp** [comment]
FWIW this looks good to me, but I'll leave final approval to the other reviewers

**Pull request merged**